OpenSim/Actuators/VandenBogert2011Muscle.cpp

```diff
@@ -0,0 +1,1234 @@
+        double c3 = vMaxNorm * fv_AHill * (fv_maxMultiplier - 1.0) /
+                (fv_AHill + 1.0); // parameter in the eccentric f-v equation
+        double c = lambda*c3;
+        //F2 = (g_{max} * V_{fiber} + c) / (V_{fiber} + c)
+        double katzDenom = (fiberVelocityNorm  + c);
+        F2 = (fv_maxMultiplier * fiberVelocityNorm  + c) / katzDenom ;
+        if (returnJacobians) {
+            dF2_dfiberVelocityNorm = (fv_maxMultiplier - F2) / katzDenom ;
+            dF2_dactiv = dlambda_da * c3 * fiberVelocityNorm *
+                    (1-fv_maxMultiplier) / pow(katzDenom,2);
+        }
+    }
+    if (returnJacobians){
+        dF2_dprojFibVelNorm =
+                dF2_dfiberVelocityNorm * dfiberVelocityNorm_dprojFibVelNorm;
+        dF2_dprojFibLenNorm =
+                dF2_dfiberVelocityNorm * dfiberVelocityNorm_dprojFibLenNorm;
+    }
+
+
+    //------F3 is the dimensionless fiber (PEE) force (in units of Fmax)------//
+
+
+
+    double dF3_dprojFibLenNorm;
+
+    // stiffness of the linear term is 1 N/m, convert to Fmax/Lceopt units
+    double kPEE = 1.0 / maxIsoForce * optFiberLength;
+    // elongation of fiber (PEE), relative to Lceopt
+    double kPEE2Norm = 1 / pow(fl_width, 2);  //Fiber (PEE) quadratic stiffness,
+    //          so Fpee = Fmax when Lce = Lce*(1+W)
+    double elongationFiberNorm = (fiberLengthNorm - fiberSlackLengthNorm);
+    double F3 = kPEE * elongationFiberNorm;
+    // low stiffness linear term
+    double dF3_dfiberLengthNorm = kPEE;
+    if (elongationFiberNorm > 0) {
+        //add quadratic term for positive elongation
+        F3 = F3 + kPEE2Norm * pow(elongationFiberNorm, 2);
+        if (returnJacobians) {
+            dF3_dfiberLengthNorm =
+                    dF3_dfiberLengthNorm + 2 * kPEE2Norm * elongationFiberNorm;
+        }
+    }
+    if (returnJacobians) {
+        dF3_dprojFibLenNorm = dF3_dfiberLengthNorm * dfiberLength_dprojFibLen;
+    }
+
+
+    //--------F4 is the dimensionless SEE force (in units of Fmax)----------//
+
+
+    // stiffness of the linear term is 1 N/m, convert to Fmax/m (so normalized
+    // by Fmax)
+    double kSEE = 1.0 / maxIsoForce;
+
+    // Tendon (SEE) quadratic stiffness, so Fsee = Fmax at strain of umax
+    // This is normalized by Fmax
+    double kSEE2 = maxIsoForce / pow(tendonSlackLength * fMaxTendonStrain, 2);
+
+    // elongation of tendon (SEE), in meters
+    double elongationTendon = muscleLength - projFibLenNorm *
+                                             optFiberLength - tendonSlackLength;
+    //  low stiffness linear term
+    double F4 = kSEE * elongationTendon;
+    double dF4_dmuscleLength;
+    double dF4_dprojFibLenNorm;
+    if (returnJacobians) {
+         dF4_dprojFibLenNorm = -kSEE * optFiberLength;
+         dF4_dmuscleLength = kSEE;
+    }
+    if (elongationTendon > 0) {
+        // add quadratic term for positive deformation
+        F4 = F4 + (kSEE2 * pow(elongationTendon, 2))/maxIsoForce;
```

Comment:
For the above, do you think it makes sense to create a separate function that describes the dimensionless (intrinsic) tendon force-length curve, and that does not depend on the (extrinsic) muscle properties like max isometric force, etc.?

I feel that `kSEE` need not be adjustable by users but users will likely want to alter the stiffness of the tendon, perhaps via some part of `kSEE2`.